DependencyInjection/HttplugExtension.php

```diff
@@ -359,94 +356,26 @@ private function createUri(ContainerBuilder $container, $serviceId, $uri)
     private function configureAutoDiscoveryClients(ContainerBuilder $container, array $config)
     {
         $httpClient = $config['discovery']['client'];
-
-        if (!empty($httpClient)) {
-            if ($httpClient === 'auto') {
-                $httpClient = $this->registerAutoDiscoverableClient(
-                    $container,
-                    'auto_discovered_client',
-                    $this->configureAutoDiscoveryFactory(
-                        $container,
-                        HttpClientDiscovery::class,
-                        'auto_discovered_client',
-                        $config
-                    ),
-                    $this->isConfigEnabled($container, $config['profiling'])
-                );
-            }
-
-            $httpClient = new Reference($httpClient);
+        if (!empty($httpClient) && $httpClient !== 'auto') {
+            $container->setAlias('httplug.auto_discovery.auto_discovered_client', $httpClient);
+        } elseif (null === $httpClient) {
+            $container->removeDefinition('httplug.auto_discovery.auto_discovered_client');
+            $container->removeDefinition('httplug.collector.auto_discovered_client');
         }
 
         $asyncHttpClient = $config['discovery']['async_client'];
-
-        if (!empty($asyncHttpClient)) {
-            if ($asyncHttpClient === 'auto') {
-                $asyncHttpClient = $this->registerAutoDiscoverableClient(
-                    $container,
-                    'auto_discovered_async',
-                    $this->configureAutoDiscoveryFactory(
-                        $container,
-                        HttpAsyncClientDiscovery::class,
-                        'auto_discovered_async',
-                        $config
-                    ),
-                    $this->isConfigEnabled($container, $config['profiling'])
-                );
-            }
-
-            $asyncHttpClient = new Reference($asyncHttpClient);
+        if (!empty($asyncHttpClient) && $asyncHttpClient !== 'auto') {
+            $container->setAlias('httplug.auto_discovery.auto_discovered_async', $asyncHttpClient);
+        } elseif (null === $asyncHttpClient) {
+            $container->removeDefinition('httplug.auto_discovery.auto_discovered_async');
+            $container->removeDefinition('httplug.collector.auto_discovered_async');
         }
 
         if (null === $httpClient && null === $asyncHttpClient) {
```

Comment:
Yes, If we have no discovered client (sync and async), we remove the useless listener.